abi.md

```diff
@@ -36,30 +36,69 @@ may evolve as new FIRRTL constructs are added.
 
 ## On Modules
 
-### The Circuit and Top Level
+### The Circuit
 
-The top level module, as specified by the circuit name, shall be present as a
-System Verilog module of the same name.  This FIRRTL module is considered a
-"public" module and subject to the lowering constraints for public modules.
+The circuit is a container of public modules.  The circuit, by itself, does not
+have any ABI.
 
 ### External Modules
 
 An external module may be presumed to exist following the lowering constraints
-for public modules.  The module shall exist with a verilog name matching the
-defname value or, lacking that, the module name.
+for public modules.  The module shall exist with a Verilog name matching the
+defname value or, lacking that, the FIRRTL module name.  The ports of an
+external module shall adhere to one of the port lowering ABIs.
 
-###  Public Modules
+No assumption is made of the filename of an implementation of an external
+module.  It is the user's job to include files in such a way in their tools to
+resolve the name.
 
-Any module considered a "public" module shall be implemented in Verilog in a
-consistent way.  Any public module shall exist as a Verilog module of the same
-name.
+### Public Modules
 
-Each public module with definitions (e.g. not external modules) shall be placed
-in a file with the same name.  No assumption is made of the filename of an
-implementation of an external module, it is the user's job to include files in
-such a way in their tools to resolve the name.
+Any public modules shall be implemented in Verilog in a consistent way.  Any
+public module shall exist as a Verilog module of the same name.  Each public
+module shall be placed in a file with a format as follows where `module` is the
+name of the public module:
 
-### Port Lowering ABIv1
+``` ebnf
+filename = module , ".sv" ;
+```
+
+Each public module in a circuit shall produce a filelist that contains the
+filename of the file containing the public module and any necessary files that
+define all public or private module files instantiated under it.  Files that
+define external modules are not included.  This filelist shall have a name as
+follows where `module` is the name of the public module:
+
+``` ebnf
+filelist_filename = "filelist_" , module , ".f" ;
+```
+
+The filelist contents shall include the name of the file containing the public
```

Comment:
does file order matter?  Are there tools that would require modules appear before instantiations?  Is this over specifying?

Reply:
Yes, this is a weird thing to bake into the ABI. I'll change this to say that this is a newline delimited list of filenames.

Reply:
Changes in 996e307.